### PR TITLE
Update the block theme folders to templates and parts

### PR DESCRIPTION
**Pull request by youknowriad**
See #36548

This PR updates the block themes to use `templates` and `parts` instead of `block-templates` and `block-template-parts`. 
It also keeps backward compatibility, that said a theme can not have "templates" and "block-template-parts" or "block-templates" and "parts". I also updated all the documentation we have.

**Notes**

The issue also mentions the future of patterns and styles, I think these are not requirements for 5.9

**Testing instructions**

 - Make sure themes with "block-templates" and "block-template-parts" folders work properly.
 - Make sure themes with "templates" and "parts" folders work properly.

### CODE REVIEW

**Comment by carolinan** [comment]
Theme Check also needs to be updated, otherwise themes will not be able to be uploaded to WordPress.org
https://github.com/WordPress/theme-check/blob/master/checks/class-fse-required-files-check.php

**Comment by youknowriad** [comment]
@ockham Do you think you can help with the php unit failures here? I'm not sure I understand the results I'm having and the previous assertions.

File: lib/compat/wordpress-5.9/block-template-utils.php [hunk 1]
```diff
@@ -21,6 +21,36 @@
...
+	/**
+	 * For backward compatibility reasons,
+	 * block themes might be using block-templates or block-template-parts,
+	 * this function ensures we fallback to these folders properly.
+	 *
+	 * @param string $theme_stylesheet The stylesheet. Default is to leverage the main theme root.
+	 *
+	 * @return array Folder names used by block themes.
+	 */
+	function get_block_theme_folders( $theme_stylesheet = null ) {
```
Comment (Mamaduka) [inline lib/compat/wordpress-5.9/block-template-utils.php:35]: Why not accept `theme_dir` directly and remove calculation logic?

Comment (youknowriad) [reply lib/compat/wordpress-5.9/block-template-utils.php:35]: From an API point of view, I felt that accepting a them stylesheet (or id) is a bit better.

Comment (Mamaduka) [reply lib/compat/wordpress-5.9/block-template-utils.php:35]: No strong preference. I was just curious :)

**Comment by ockham** [comment]
> @ockham Do you think you can help with the php unit failures here? I'm not sure I understand the results I'm having and the previous assertions.

@youknowriad I'll have a look!

**Comment by ockham** [comment]
All of the `*_block_template_takes_precedence_over_*` tests are failing, which means that the template resolution algorithm thinks that the PHP templates always take precedence -- probably because it can't locate the block templates. So something in template resolution seems to be a bit off still...

**Comment by ockham** [comment]
Is it possible that the reason for this issue is that the test (end dev) env [is using Core `trunk`](https://github.com/WordPress/gutenberg/blob/4d1b2dcbf5b3ba78fd58c18c4a4b17e6c7b0011f/.wp-env.json#L2), and thus the original versions of the template resolution algorithm that we already backported (since it finds those functions and doesn't need GB's 5.9 compat layer)?

The reason why I think this might be the case is as follows: If I move the text fixture files back from `phpunit/fixtures/themes/test-theme/templates/` to `phpunit/fixtures/themes/test-theme/block-templates/`, the lines that assert that the template resolution algorithm finds the block template no longer fail. In other words, the template resolution algorithm expects to find the templates in `block-templates/` (rather than `templates/`).

**Comment by ockham** [comment]
Specifically, I think that GB is [picking](https://github.com/WordPress/gutenberg/blob/8973e2c5a5979535c321e73fa12a8690b80a3841/lib/compat/wordpress-5.9/block-template-utils.php#L208) [up](https://github.com/WordPress/gutenberg/blob/8973e2c5a5979535c321e73fa12a8690b80a3841/lib/compat/wordpress-5.9/block-template-utils.php#L259) Core's versions of `_get_block_template_file` and `_get_block_templates_files`, respectively, which are hard-wired to `block-templates` and `block-template-parts`.

**Comment by youknowriad** [comment]
@ockham that makes sense, I guess we need to make the fix in both places at the same time.

**Comment by youknowriad** [comment]
I think actually that we should remove the unit test from here, now that these functions are just in Gutenberg to support old versions.

**Comment by ockham** [comment]
> I think actually that we should remove the unit test from here, now that these functions are just in Gutenberg to support old versions.

Let's just make sure that we've carried them over to `wordpress-develop` so we don't lose test coverage for some of these cases 👍

**Comment by ockham** [comment]
> > I think actually that we should remove the unit test from here, now that these functions are just in Gutenberg to support old versions.
> 
> Let's just make sure that we've carried them over to `wordpress-develop` so we don't lose test coverage for some of these cases 👍

Looks like we still have to backport them: https://github.com/WordPress/wordpress-develop/blob/d87196b560203345b385bef6bef9666e7147bb46/tests/phpunit/tests/block-template.php

**Comment by ockham** [comment]
> > > I think actually that we should remove the unit test from here, now that these functions are just in Gutenberg to support old versions.
> > 
> > 
> > Let's just make sure that we've carried them over to `wordpress-develop` so we don't lose test coverage for some of these cases 👍
> 
> Looks like we still have to backport them: https://github.com/WordPress/wordpress-develop/blob/d87196b560203345b385bef6bef9666e7147bb46/tests/phpunit/tests/block-template.php

I've started a PR: https://github.com/WordPress/wordpress-develop/pull/1920

**Comment by ockham** [comment]
> I think actually that we should remove the unit test from here, now that these functions are just in Gutenberg to support old versions.

https://github.com/WordPress/gutenberg/pull/36855

**Comment by youknowriad** [comment]
The tests here are passing, so this should be good to land now.

**Review by aristath — APPROVED** [review]

**Pull request merged**